async_js.md

```diff
@@ -0,0 +1,289 @@
+
+Endi bir o'ylab ko'ringchi console'ga qanday natija chiqadi.
+
+Agar sizning natijangiz quyidagicha bo'lsa, siz ko'pchilik kabi xato o'ylagansiz va bu tabiiy holat.
+
+```
+1 - chiqadi
+2 - chiqadi
+3 - chiqadi
+```
+
+To'g'ri javob quyidagicha bo'ladi
+
+```
+1 - chiqadi
+3 - chiqadi
+2 - chiqadi
+```
+
+Siz aytishingiz mumkin `setTimeout()` funksiyaga vaqt sifatida nol berilganku. Bu degani bu funksiya hech qancha vaqt kutmasdan o'zini vazifasini bajarishi kerak deb. Ha aslida shunday. Lekin, bu yerda hammma gap **sinxron** va **asinxron** funksiyalarda. [`console.log()`](https://developer.mozilla.org/en-US/docs/Web/API/console/log) - sinxron, `setTimeout()` - asinxron funksiya hisoblanadi. Va asinxron funksiya sinxron funksiyadan so'ng bajariladi.
+
+JavaScript Engine kodni o'qish jarayonida asinxron funksiyalarni [queue](<https://dmitripavlutin.com/javascript-queue/#:~:text=The%20queue%20data%20structure%20is,constant%20time%20O(1)%20.>)(navbat)ga yig'ib boradi va hamma sinxron funksiyalarni bararib bo'lgandan so'ng, navbati bilan queue'dagi asinxron funksiyalarni bajarishni boshlaydi.
+
+Shuning uchun bizni misolimizda setTimeout funksiyaga nol argumenti berilgani bilan ham u oxirida bajarilyapti.
+
+Quyidagi misolni o'zingiz tahlil qilib ko'ring va browser console'ida o'zingizni tekshirib ko'ring.
+
+```js
+console.log("1 - chiqadi");
+setTimeout(() => console.log("2 - chiqadi"), 10);
+setTimeout(() => console.log("3 - chiqadi"), 0);
+console.log("4 - chiqadi");
+```
+
+## Promise
+
+Biz yuqorida sinxron va asinxron funksiyalar bilan tanishdik. Endi **asinxron** funksiya ishlash prinsipi asosida qurilgan metodlardan biri [**Promise**](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) lar bilan tanishamiz.
+
+Tasavvur qiling siz media fayllarga boy biror saytga tashrif buyurayapsiz. Sizda u sayt quyidagi ikki xil ko'rinishda yuklanishi mumkin. **1 - ko'rinishi:** saytdagi media fayllar va context(matn)lar birinma - ketin yuklanishi yoki **2 - ko'rinishi:** oldin contextlar yuklanib keyin media fayllar yuklanishi mumkin. Bunda 1 - holatda sayt o'ziga ko'p foydalanuvchilarni jalb qila olmasligi mumkin. Sababi bunday saytlarda qotishlar bo'lishi mumkin. Chunki kattaroq media fayllarni browserga joylashtirish ko'proq vaqt oladi. Tabiiyki, foydalanuvchi noma'lum sayt yuklanishini uzoq vaqt kutishni yoqtirmaydi. 2 - holatda esa saytdagi matnli ma'lumotlar birinchi to'liq yuklanadi, so'ng media fayllar. Bu degani saytni to'liq yuklanishini kutmasdan turib, saytdagi mantlardan bu sayt sizga kerakli yoki yo'qligini aniqlashingiz mumkin degani. Xulosa qiladigan bo'lsak, siz saytingizni 2 - usul orqali yaratsangiz, sizning sayt birinchisiga nisbatan foydalanuvchiga qulay bo'ladi. Bunday saytlarni yaratishda esa siz _Promise_'lardan foydalanishingiz mumkin.
```

Comment:
Husan keyin qarang bu yerda ikki xil yulanishi mumkin degan gapdan keyin yangi qatorga tushadimi yoki gapning davomida bog'lanib ketganmi,  keyin 2 xil ko'rinish deb yozgansiz shundan keyin yana alohida-alohida 1-ko'rinish va 2-ko'rinish deb yozipsiz menimcha bu shartmas.   1.    va 2. shunaqa qilib yangi qatorlardan yozib qo'ysangiz yaxshiroq ko'rinadi meni nazarimda

Reply:
To'g'ri fikr. To'g'irlaymi buni

Reply:
> Bu degani saytni to'liq yuklanishini kutmasdan turib, saytdagi ~~mantlardan~~ -> matnlardan

Reply:
Promise haqida shu joyiga yana qandaydir hayotiy misolga o'xshatib tushuntirish kerak, menimcha. Masalan, 

> Restoranga borib ovqat buyurtma qilasiz. Ovqat tayyor bo'lganida sizga keltirishadi. Siz ungacha qo'lingizni yuvib kelishingiz yoki o'rtog'ingiz bilan gaplashib turishingiz mumkin. Bir ishning bitishini kutmasdan boshqa(lar)ini bajarish asinxron ish bajarish deyiladi. 
Endi tasavvur qiling, siz ovqat tayyor bo'lib olib kelishmagunicha hech narsa qilmasdan kutib o'tirishingiz kerak bo'lsa, unda siz ishni ketma-ket yoki sinxron bajargan bo'lasiz. Ammo bu juda sekin va zerikarli kechadi, to'g'rimi? 
Ofitsiantga ovqat buyurishga o'xshash jarayonni Javascript'da Promise bajaradi.